### ISSUE TITLE
SCIP 8.0 support

### ISSUE THREAD

**Issue by chemrgineer** [Pyomo/pyomo#2226]
## Summary

Pyomo solver interface works with SCIP 7 via scipampl reader extention. New verison of SCIP integrated AMPL reader so Pyomo need to be updated. 

### Description

I'm using pyomo and SCIP for MINLP problems. Recently SCIP updated as a beta version but it works very well. I tried to run with SCIP 8 with solver executable option but not worked. 
**Example code:**
from pyomo.environ import *
m = ConcreteModel()
m.x1 = Var(bounds=(0,6))
m.x2 = Var(bounds=(0,4))
m.cons1 = Constraint(expr= (m.x1)*(m.x2) <= 4)
m.obj = Objective(expr= -(m.x1+m.x2))
solver = SolverFactory('scip', executable="/home/optimization/scip/build/bin/scip")
result = solver.solve(m, tee=True)

**Terminal output:**
```
SCIP version 8.0.0.1 [precision: 8 byte] [memory: block] [mode: optimized] [LP solver: CPLEX 20.1.0.0] [GitHash: f8ca16eaa1]
Copyright (C) 2002-2021 Konrad-Zuse-Zentrum fuer Informationstechnik Berlin (ZIB)

External libraries: 
  Readline 8.1         GNU library for command line editing (gnu.org/s/readline)
  CPLEX 20.1.0.0       Linear Programming Solver developed by IBM (www.cplex.com)
  CppAD 20180000.0     Algorithmic Differentiation of C++ algorithms developed by B. Bell (github.com/coin-or/CppAD)
  ZLIB 1.2.11          General purpose compression library by J. Gailly and M. Adler (zlib.net)
  GMP 6.0.0            GNU Multiple Precision Arithmetic Library developed by T. Granlund (gmplib.org)
  AMPL/MP 4e2d45c4     AMPL .nl file reader library (github.com/ampl/mp)
  PaPILO 2.0.0         parallel presolve for integer and linear optimization (github.com/scipopt/papilo) [GitHash: 54d2a71]
  bliss 0.73           Computing Graph Automorphism Groups by T. Junttila and P. Kaski (www.tcs.hut.fi/Software/bliss/)
  Ipopt 3.14.4         Interior Point Optimizer developed by A. Waechter et.al. (github.com/coin-or/Ipopt)

user parameter file <scip.set> not found - using default parameters

read problem </tmp/tmppvdng4cv.pyomo.nl.nl>
============

[reader_nl.cpp:1351] ERROR: cannot open file /tmp/tmppvdng4cv.pyomo.nl.nl: No such file or directory
[scipshell.c:190] ERROR: Error <-4> in function call
[scipshell.c:246] ERROR: Error <-4> in function call
[scipshell.c:533] ERROR: Error <-4> in function call
SCIP Error (-4): file not found error
ERROR: Solver (scip) returned non-zero return code (255)
ERROR: See the solver log above for diagnostic information.
Traceback (most recent call last):
  File "/home/optimization/pyomo_optimization/minlp1.py", line 15, in <module>
    result = solver.solve(m, tee=True)
  File "/home/optimization/miniconda3/lib/python3.8/site-packages/pyomo/opt/base/solvers.py", line 595, in solve
    raise ApplicationError(
pyomo.common.errors.ApplicationError: Solver (scip) did not exit normally
```

In addition to this update, SCIP can work parallel with "concurrentopt" command. Is it possibe to determine from pyomo?
Sorry for my mistakes. Thanks for your answers.

**Comment by chemrgineer** [comment]
I thing solver interface will not updated soon. So Is there anyone can help me to handle this problem. I tried newly compiled scip executable rename "scipampl" to trick pyomo but its not helpful. I think key point is pyomo can not pass .nl file properly.
"[reader_nl.cpp:1351] ERROR: cannot open file /tmp/tmppvdng4cv.pyomo.nl.nl: No such file or directory
[scipshell.c:190] ERROR: Error <-4> in function call
[scipshell.c:246] ERROR: Error <-4> in function call
[scipshell.c:533] ERROR: Error <-4> in function call
SCIP Error (-4): file not found error"

**Comment by vanamalivanam** [comment]
i have had a hack by renaming the files in shellcmd.py of pyomo library. not very clean but definitely works with scip8 directly. path to pyomo script is here
/<venv>/lib/python3.9/site-packages/pyomo/opt/solver/shellcmd.py

**Comment by vanamalivanam** [comment]
@chemrgineer I tested the changes on both pyomo 6.4 and pyomo 6.2 versions with scip8, and it works.
the changes below are made on pyomo 6.4 version. 
the change below is in _apply_solver(self) method: 
<img width="1039" alt="image" src="https://user-images.githubusercontent.com/7015067/159002195-4de58f0c-6c0a-4a85-8c15-d22537fa2339.png">

the change below is in process_output() method:
<img width="917" alt="image" src="https://user-images.githubusercontent.com/7015067/159002753-c4b2a82d-2783-4a9a-aee5-2f190a5d4411.png">

**Comment by chemrgineer** [comment]
@vanamalivanam I tried your solution but I could not succeed. I tested with Pyomo 6.4 and SCIP 8.0 release version.

**Comment by milannnnn** [comment]
@chemrgineer I ran into the same issue and realized that I was not using the `keepfiles` flag (due to which the renaming never took place). Here's the modification that should make it work in both cases:

```
    def _apply_solver(self):
        print('Applying Solver')
        if pyomo.common.Executable('timer'):
            self._timer = pyomo.common.Executable('timer').path()
        #
        # Execute the command
        #
        if is_debug_set(logger):
            logger.debug("Running %s", self._command.cmd)

        if self._problem_files is not []:
            newtup = ()
            for pfilestr in self._problem_files:
                newfname = pfilestr.replace('.nl', '.nl.nl')
                os.rename(pfilestr, newfname)
                newtup = newtup + (newfname,)
            self._problem_files = newtup

        # display the log/solver file names prior to execution. this is useful
        # in case something crashes unexpectedly, which is not without precedent.
        if self._keepfiles:
            if self._log_file is not None:
                print("Solver log file: '%s'" % self._log_file)
            if self._soln_file is not None:
                print("Solver solution file: '%s'" % self._soln_file)
            if self._problem_files is not []:
                print("Solver problem files: %s" % str(self._problem_files))

        sys.stdout.flush()
        self._rc, self._log = self._execute_command(self._command)
        sys.stdout.flush()
        return Bunch(rc=self._rc, log=self._log)
```
and
```
    def process_output(self, rc):
        """
        Process the output files.
        """
        start_time = time.time()
        if self._results_format is None:
            raise ValueError("Results format is None")
        results = self.process_logfile()
        log_file_completion_time = time.time()
        if self._report_timing is True:
            print("      %6.2f seconds required to read logfile " % (log_file_completion_time - start_time))
        if self._results_reader is None:
            self.process_soln_file(results)
            soln_file_completion_time = time.time()
            if self._report_timing is True:
                print("      %6.2f seconds required to read solution file " % (soln_file_completion_time - log_file_completion_time))
        else:
            # There is some ambiguity here as to where the solution data
            # It's natural to expect that the log file contains solution
            # information, but perhaps also in a results file.
            # For now, if there is a single solution, then we assume that
            # the results file is going to add more data to it.
            if len(results.solution) == 1:
                results = self._results_reader(self._results_file,
                                               res=results,
                                               soln=results.solution(0),
                                               suffixes=self._suffixes)
            else:
                oldname = self._results_file
                newname = oldname.replace('.sol', '.nl.sol')
                self._results_file = newname
                results = self._results_reader(self._results_file,
                                               res=results,
                                               suffixes=self._suffixes)
            results_reader_completion_time = time.time()
            if self._report_timing is True:
                print("      %6.2f seconds required to read solution file" % (results_reader_completion_time - log_file_completion_time))
        if rc != None:
            results.solver.error_rc=rc
            if rc != 0:
                results.solver.status=SolverStatus.error

        if self._last_solve_time != None:
            results.solver.time=self._last_solve_time

        return results
```

**Comment by chemrgineer** [comment]
@milannnnn and @vanamalivanam thank you for your help. I successfuly passed my problem to SCIP 8.0 on Windows 10. I will compile latest source code and try on linux. I hope, main pyomo will implament your codes as soon as possible.

**Comment by sebuer** [comment]
I ran into the same issue mentioned here and worked on a fix during the same time.

Instead of renaming the problem file as mentioned by @vanamalivanam, I chose the route of changing the command line arguments depending on the solver version. There is also no need to rename the solver to `scipampl`. So I would assume, that the changes you made break compatibility to SCIP < 8. Did you by any chance test this compatibility?

I also noticed, that the `scip` executable ignores the solver settings in combination with the `-AMPL` argument. So maybe you would also get the output
`user parameter file <scip.set> not found - using default parameters`
as I did, testing these changes.

@chemrgineer, I did not test my changes in the pull request under windows.

I would be happy for any feedback regarding the pull request, if anyone notices something.
